Dev implement grafana auth interface

saltiyazan opened the pull request:
- Implements grafana-auth-interface
- Adds unit test cases
- Adds a dummy charm
- Adds description and README

saltiyazan committed:
Adds dummy charm for the lib

saltiyazan committed:
Implements grafana-auth interface

saltiyazan committed:
Adds github workflow and tox

saltiyazan committed:
Adds unit testing for the lib

saltiyazan committed:
Adds requirements file

saltiyazan committed:
Adds content to README.md

saltiyazan committed:
Fixes typos

ghislainbourgeois commented on README.md:
```diff
@@ -0,0 +1,22 @@
+
+This project contains libraries for the grafana-auth relationship. It contains both the provider and the requirer side of them.
+
+> Note: The charm located here is a placeholder charm and should not packed nor should be deployed.
```
```suggestion
> Note: The charm located here is a placeholder charm and should not be packed nor  deployed.
```

saltiyazan replied:
Done

ghislainbourgeois commented on lib/charms/grafana_auth_interface/v0/grafana_auth_interface.py:
```diff
@@ -0,0 +1,456 @@
+        if not current_auth_conf:
+            logger.info("No authentication configuration found")
+            return
+        unrequired_auth_modes = list(current_auth_conf.get("auth").keys())
```
I think it's not really `unrequired` in that case, but more `unprovided`, no?

saltiyazan replied:
I changed it to `revoked`, what do you think of that?
I'm also okay with `unprovided `

simonaronsson replied:
I definitely prefer `unprovided` over `unrequired` and `revoked`. It's not that we've ruled them out, just that we haven't provided them through this particular relation.

ghislainbourgeois replied:
I also prefer `unprovided`, as I think it is quite explicit in what happened, the authentication mode stopped being provided. `revoked` to me implies some intent to block access.

saltiyazan replied:
@gruyaume Suggested to remove it, and let the charm respond to the `relation_departed` event and do what it needs to do.
Do you have any thoughts on that? I will change it to `unprovided` in case it stays :)

ghislainbourgeois replied:
I think I agree with @gruyaume, this will be less code to maintain in this library.

saltiyazan replied:
Removed

ghislainbourgeois commented on requirements.txt:
```diff
@@ -0,0 +1,2 @@
+git+https://github.com/canonical/operator/#egg=ops
```
I think we now want to target `ops` directly

saltiyazan replied:
Done

saltiyazan committed:
Adds missing python docstring

ghislainbourgeois commented on tests/charms/grafana_auth_interface/test_charm.py:
```diff
@@ -0,0 +1,320 @@
+        f"{CHARM_LIB_PATH}.GrafanaAuthRequirerCharmEvents.grafana_auth_config_available",
+        new_callable=PropertyMock,
+    )
+    def test_given_grafana_auth_conf_in_relation_databag_when_unit_is_leader_and_schema_is_valid_then_grafana_auth_config_available_event_is_emitted(
```
```suggestion
    def test_given_grafana_auth_conf_in_relation_databag_and_unit_is_leader_and_schema_is_valid_when_grafana_auth_relation_changed_then_grafana_auth_config_available_event_is_emitted(
```

saltiyazan replied:
Done

ghislainbourgeois commented on tests/charms/grafana_auth_interface/test_charm.py:
```diff
@@ -0,0 +1,320 @@
+        f"{CHARM_LIB_PATH}.GrafanaAuthRequirerCharmEvents.grafana_auth_config_available",
+        new_callable=PropertyMock,
+    )
+    def test_given_grafana_auth_conf_in_relation_databag_when_unit_is_not_leader_then_grafana_auth_config_available_event_not_is_emitted(
```
```suggestion
    def test_given_grafana_auth_conf_in_relation_databag_and_unit_is_not_leader_when_grafana_auth_relation_changed_then_grafana_auth_config_available_event_not_is_emitted(
```

saltiyazan replied:
Done

ghislainbourgeois commented on tests/charms/grafana_auth_interface/test_charm.py:
```diff
@@ -0,0 +1,320 @@
+        f"{CHARM_LIB_PATH}.GrafanaAuthRequirerCharmEvents.grafana_auth_config_available",
+        new_callable=PropertyMock,
+    )
+    def test_given_grafana_auth_conf_in_relation_databag_when_auth_mode_not_valid_then_grafana_auth_config_available_event_is_not_emitted(
```
```suggestion
    def test_given_grafana_auth_conf_in_relation_databag_and_auth_mode_not_valid_when_grafana_auth_relation_changed_then_grafana_auth_config_available_event_is_not_emitted(
```

saltiyazan replied:
Done

ghislainbourgeois commented on tests/charms/grafana_auth_interface/test_charm.py:
```diff
@@ -0,0 +1,320 @@
+        f"{CHARM_LIB_PATH}.GrafanaAuthRequirerCharmEvents.grafana_auth_config_available",
+        new_callable=PropertyMock,
+    )
+    def test_given_grafana_auth_conf_in_relation_databag_when_schema_is_not_valid_then_grafana_auth_config_available_event_is_not_emitted(
```
```suggestion
    def test_given_grafana_auth_conf_in_relation_databag_and_schema_is_not_valid_when_grafana_auth_relation_changed_then_grafana_auth_config_available_event_is_not_emitted(
```

saltiyazan replied:
Done

ghislainbourgeois commented on tests/charms/grafana_auth_interface/test_charm.py:
```diff
@@ -0,0 +1,320 @@
+        f"{CHARM_LIB_PATH}.GrafanaAuthRequirerCharmEvents.grafana_auth_config_unrequired",
+        new_callable=PropertyMock,
+    )
+    def test_given_grafana_auth_conf_in_relation_databag_when_unit_is_leader_and_relation_departed_then_grafana_auth_config_unrequired_is_emitted(
```
```suggestion
    def test_given_grafana_auth_conf_in_relation_databag_and_unit_is_leader_and_relation_departed_when_grafana_auth_relation_departed_then_grafana_auth_config_unrequired_is_emitted(
```

saltiyazan replied:
Done

gruyaume commented:
Can you add our team (telco) as an admin of this project?

gruyaume commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,35 @@
+name: CI
```
Also add a upload actions

saltiyazan replied:
Done

gruyaume commented on LICENSE:
```diff
@@ -0,0 +1,202 @@
+      same "printed page" as the copyright notice for easier
+      identification within third-party archives.
+
+   Copyright [yyyy] [name of copyright owner]
```
```suggestion
   Copyright 2022 Canonical Ltd.
```

saltiyazan replied:
Done

gruyaume commented on charmcraft.yaml:
```diff
@@ -0,0 +1,10 @@
+# Learn more about charmcraft.yaml configuration at:
```
Those 2 lines can be removed

saltiyazan replied:
Done

gruyaume commented on README.md:
```diff
@@ -0,0 +1,22 @@
+
+This project contains libraries for the grafana-auth relationship. It contains both the provider and the requirer side of them.
+
+> Note: The charm located here is a placeholder charm and should not packed nor should be deployed.
```
```suggestion
> Warning: The charm located here is a placeholder charm and should not packed nor should be deployed.
```

saltiyazan replied:
Done

gruyaume commented on metadata.yaml:
```diff
@@ -0,0 +1,18 @@
+# Copyright 2022 Ubuntu
+# See LICENSE file for licensing details.
+
+# For a complete list of supported options, see:
```
Those 2 lines can be removed

saltiyazan replied:
Done

gruyaume commented on README.md:
````diff
@@ -0,0 +1,22 @@
+```bash
+charmcraft fetch-lib charms.grafana_auth_interface.v0.grafana_auth_interface
+```
+
````
There should also be something about the libs necessary for those charms to add to their `requirements.txt` file.

gruyaume replied:
In this case, only `jsonschema`

saltiyazan replied:
Done

gruyaume commented on lib/charms/grafana_auth_interface/v0/grafana_auth_interface.py:
```diff
@@ -0,0 +1,476 @@
+"""grafana-auth interface library.
```
Missing Copyright line

saltiyazan replied:
Done

sed-i replied:
Copyright messages can be linted ([example](https://github.com/canonical/alertmanager-k8s-operator/blob/fb9ad0012e125eac56049efe002bcf35839db1e9/pyproject.toml#L35)).

gruyaume commented on lib/charms/grafana_auth_interface/v0/grafana_auth_interface.py:
```diff
@@ -0,0 +1,476 @@
+            return
+        self.on.grafana_auth_config_available.emit(auth_conf=auth_conf)
+
+    def _on_grafana_auth_relation_departed(self, event):
```
I don't know about this method. Maybe the better approach is the simplest. Simply remove it and have the charm author do what they want when they receive the `_grafana_auth_relation_departed`. What do you think?

saltiyazan replied:
Yes, I thought of this, then simply the charm should observe the `on_relation_departed` event and it does whatever it wants with that :)
The only idea here is to remind the charm of the auth types that were passed through the relation.
I don't insist on keeping it and can remove it as the charm itself could do the same.
What do you think?

gruyaume replied:
I think I'd rather remove it

saltiyazan replied:
Done

gruyaume commented on lib/charms/grafana_auth_interface/v0/grafana_auth_interface.py:
```diff
@@ -0,0 +1,476 @@
+        """
+        url_dict = dict()
+        url_dict["url"]= url
+        return url_dict
```
```suggestion
        return {"url": url}
```

gruyaume replied:
Actually, is this method needed at all, we could simply have `self._grafana_url = {"url": url}` in the __init__

saltiyazan replied:
I agree, Done

gruyaume commented on lib/charms/grafana_auth_interface/v0/grafana_auth_interface.py:
```diff
@@ -0,0 +1,476 @@
+            charm.on[relationship_name].relation_joined, self._set_grafana_url_in_realtion_data
+        )
+
+    def _build_url_dict(self, url: str):
```
Add return typing

saltiyazan replied:
Done

gruyaume commented on lib/charms/grafana_auth_interface/v0/grafana_auth_interface.py:
```diff
@@ -0,0 +1,476 @@
+        conf_dict["auth"]= {auth_type:{ key:value for key, value in kwargs.items()}}
+        return conf_dict
+
+    def _set_auth_config_in_realtion_data(self, event: RelationJoinedEvent) -> None:
```
```suggestion
    def _set_auth_config_in_relation_data(self, event: RelationJoinedEvent) -> None:
```

saltiyazan replied:
Done

gruyaume commented on lib/charms/grafana_auth_interface/v0/grafana_auth_interface.py:
```diff
@@ -0,0 +1,476 @@
+        unrequired_auth_modes = list(current_auth_conf.get("auth").keys())
+        self.on.grafana_auth_config_unrequired.emit(unrequired_auth_modes)
+
+    def _set_grafana_url_in_realtion_data(self, event: RelationJoinedEvent) -> None:
```
```suggestion
    def _set_grafana_url_in_relation_data(self, event: RelationJoinedEvent) -> None:
```

saltiyazan replied:
Done

saltiyazan commented:
> Can you add our team (telco) as an admin of this project?

Done :)

saltiyazan committed:
Addresses review comments
Fixes typos in README
Fixes event name in lib, changes AuthConfUnrequiredEvent to AuthConfRevokedEvent
Points directly to ops in requirements
Fixes test case names

saltiyazan committed:
Addresses review comments
Adds copyrights
Fixes typos in grafana_auth_interface.py
Remove unnecessary comment lines

saltiyazan committed:
Adds upload github action

saltiyazan committed:
Fixes typos

saltiyazan committed:
Fixes tox static issues

saltiyazan committed:
Improves docstring documentation and adds return types

simonaronsson commented:
I've yet to review the code thoroughly, but why do we need a separate repo for the library? Shouldn't this lib be living in the Grafana repo? We will need a `grafana-auth-k8s` charm at some point as we start to implement the modes that need property/behavioral mappings, but the library itself better live with the charm it supports.

@rbarry82 feel free to have opinions about whether it should be a separate repo (with a placeholder charm, like observability-libs) or not, but my vote is for merging this into grafana-k8s.

saltiyazan commented:
> I've yet to review the code thoroughly, but why do we need a separate repo for the library? Shouldn't this lib be living in the Grafana repo? We will need a `grafana-auth-k8s` charm at some point as we start to implement the modes that need property/behavioral mappings, but the library itself better live with the charm it supports.
> 
> @rbarry82 feel free to have opinions about whether it should be a separate repo (with a placeholder charm, like observability-libs) or not, but my vote is for merging this into grafana-k8s.

I'm open to moving it to live under the Grafana repo :) 
What do you think @gruyaume ?

gruyaume commented:
> > I've yet to review the code thoroughly, but why do we need a separate repo for the library? Shouldn't this lib be living in the Grafana repo? We will need a `grafana-auth-k8s` charm at some point as we start to implement the modes that need property/behavioral mappings, but the library itself better live with the charm it supports.
> > @rbarry82 feel free to have opinions about whether it should be a separate repo (with a placeholder charm, like observability-libs) or not, but my vote is for merging this into grafana-k8s.
> 
> I'm open to moving it to live under the Grafana repo :) What do you think @gruyaume ?

I have no preference, I'll leave the decision to @rbarry82 and @simskij as it's very much a grafana centric library.

saltiyazan committed:
Removes the handler for the relation_departed event

gruyaume commented on tests/charms/grafana_auth_interface/test_charm.py:
```diff
@@ -0,0 +1,292 @@
+        actual_conf_dict = _load_relation_data(event.relation.data[event.app]).get("grafana_auth")
+        self.assertDictEqual(expected_conf_dict, actual_conf_dict)  # type: ignore[arg-type]
+
+    def test_given_auth_mode_config_and_unit_is_not_leader_when_relation_joined_then_auth_conf_is_set_not_set_in_relation_databag(
```
```suggestion
    def test_given_auth_mode_config_and_unit_is_not_leader_when_relation_joined_then_auth_conf_is_not_set_in_relation_databag(
```

saltiyazan replied:
Done

gruyaume commented on tests/charms/grafana_auth_interface/test_charm.py:
```diff
@@ -0,0 +1,292 @@
+        f"{CHARM_LIB_PATH}.GrafanaAuthProviderCharmEvents.grafana_url_available",
+        new_callable=PropertyMock,
+    )
+    def test_given_grafana_grafana_url_databag_when_unit_is_leader_and_schema_is_valid_then_grafana_url_available_event_is_emitted(
```
```suggestion
    def test_given_grafana_url_in_databag_and_unit_is_leader_when_on_grafana_auth_relation_changed_then_grafana_url_available_event_is_emitted(
```

saltiyazan replied:
Done

gruyaume commented on tests/charms/grafana_auth_interface/test_charm.py:
```diff
@@ -0,0 +1,292 @@
+        f"{CHARM_LIB_PATH}.GrafanaAuthProviderCharmEvents.grafana_url_available",
+        new_callable=PropertyMock,
+    )
+    def test_given_grafana_grafana_url_databag_when_unit_is_not_leader_then_grafana_url_available_event_is_not_emitted(
```
```suggestion
    def test_given_grafana_url_not_in_databag_and_unit_is_not_leader_when_on_grafana_auth_relation_changed_then_grafana_url_available_event_is_not_emitted(
```

saltiyazan replied:
Done

gruyaume commented on tests/charms/grafana_auth_interface/test_charm.py:
```diff
@@ -0,0 +1,292 @@
+        }
+        event.app = self.requirer_app
+        self.grafana_auth_provides._on_grafana_auth_relation_changed(event)
+        patch_emit.assert_called_once()
```
We could also validate the event data `event.grafana_url`.

saltiyazan replied:
Done

gruyaume commented on tests/charms/grafana_auth_interface/test_charm.py:
```diff
@@ -0,0 +1,292 @@
+        event = Mock()
+        self.grafana_auth_provides._on_grafana_auth_relation_changed(event)
+        patch_emit.assert_not_called()
+
```
Here we have a test for 
- unit is leader + grafana url in relation databag
- unit is not leader + grafana url not in relation databag

Should we also care for the 2 other possibilities?
- unit is not leader + grafana url in relation databag
- unit is leader + grafana url not in relation databag

saltiyazan replied:
Done

gruyaume commented on tests/charms/grafana_auth_interface/test_charm.py:
```diff
@@ -0,0 +1,292 @@
+    def test_given_grafana_url_and_unit_is_leader_when_relation_joined_then_grafana_url_is_set_in_relation_databag(
+        self,
+    ):
+        expected_grafana_url = {"url": GRAFANA_URL}
```
Shouldn't we set that the unit is leader?

saltiyazan replied:
Done

gruyaume commented on tests/charms/grafana_auth_interface/test_charm.py:
```diff
@@ -0,0 +1,292 @@
+            self.charm, relationship_name=RELATIONSHIP_NAME, grafana_url=GRAFANA_URL
+        )
+
+    def test_given_grafana_url_and_unit_is_leader_when_relation_joined_then_grafana_url_is_set_in_relation_databag(
```
```suggestion
    def test_given_unit_is_leader_when_relation_joined_then_grafana_url_is_set_in_relation_databag(
```

saltiyazan replied:
Done

gruyaume commented on tests/charms/grafana_auth_interface/test_charm.py:
```diff
@@ -0,0 +1,292 @@
+        actual_grafana_url = _load_relation_data(event.relation.data[event.app]).get("grafana_url")
+        self.assertDictEqual(expected_grafana_url, actual_grafana_url)  # type: ignore[arg-type]
+
+    def test_given_grafana_url_and_unit_is_not_leader_when_relation_joined_then_grafana_url_is_not_set_in_relation_databag(
```
```suggestion
    def test_given_unit_is_not_leader_when_relation_joined_then_grafana_url_is_not_set_in_relation_databag(
```

saltiyazan replied:
Done

gruyaume commented on tests/charms/grafana_auth_interface/test_charm.py:
```diff
@@ -0,0 +1,292 @@
+        f"{CHARM_LIB_PATH}.GrafanaAuthRequirerCharmEvents.grafana_auth_config_available",
+        new_callable=PropertyMock,
+    )
+    def test_given_grafana_auth_conf_in_relation_databag_and_unit_is_leader_and_schema_is_valid_when_grafana_auth_relation_changed_then_grafana_auth_config_available_event_is_emitted(
```
```suggestion
    def test_given_grafana_auth_conf_in_relation_databag_and_unit_is_leader_when_grafana_auth_relation_changed_then_grafana_auth_config_available_event_is_emitted(
```

saltiyazan replied:
Done

gruyaume commented on tests/charms/grafana_auth_interface/test_charm.py:
```diff
@@ -0,0 +1,292 @@
+        }
+        event.app = self.provider_app
+        self.grafana_auth_requires._on_grafana_auth_relation_changed(event)
+        patch_emit.assert_called_once()
```
We should validate the event data as well

saltiyazan replied:
Done

gruyaume commented on tests/charms/grafana_auth_interface/test_charm.py:
```diff
@@ -0,0 +1,292 @@
+        self, patch_emit
+    ):
+        self.requirer_unit.set_leader(False)
+        event = Mock()
```
We should set the event relation data to `{
            self.provider_app: {"grafana_auth": json.dumps(conf_dict)},
        }`
(if we want to match with the test name)

saltiyazan replied:
Done

gruyaume commented on tests/charms/grafana_auth_interface/test_charm.py:
```diff
@@ -0,0 +1,292 @@
+    def test_given_grafana_auth_conf_in_relation_databag_and_auth_mode_not_valid_when_grafana_auth_relation_changed_then_grafana_auth_config_available_event_is_not_emitted(
+        self, patch_emit
+    ):
+        patch_emit.emit.return_value = "whatever"
```
No need for this return value

saltiyazan replied:
Done

gruyaume commented on tests/charms/grafana_auth_interface/test_charm.py:
```diff
@@ -0,0 +1,292 @@
+    def test_given_grafana_auth_conf_in_relation_databag_and_schema_is_not_valid_when_grafana_auth_relation_changed_then_grafana_auth_config_available_event_is_not_emitted(
+        self, patch_emit
+    ):
+        patch_emit.emit.return_value = "whatever"
```
No need for this return value

saltiyazan replied:
Done

saltiyazan committed:
Addresses review comments
Fixes test case names

saltiyazan committed:
Addresses review comments
Asserts event data for events that are emitted

saltiyazan committed:
Addresses review comments
Removes return_value mocks that are not needed

saltiyazan committed:
Addresses review comments
Adds new test cases

saltiyazan committed:
Changes the json schemas of the provider and the requirer

saltiyazan committed:
Fixes setting relation data to the correct databag
Fixes related unit tests

rbarry82 commented on lib/charms/grafana_auth_interface/v0/grafana_auth_interface.py:
```diff
@@ -0,0 +1,491 @@
+    def __init__(self, charm: CharmBase, relationship_name: str, auth_type: str, **kwargs):
+        super().__init__(charm, relationship_name)
+        self._charm = charm
+        self._relationship_name = relationship_name
```
We should really try to pick some default value here.

rbarry82 commented on lib/charms/grafana_auth_interface/v0/grafana_auth_interface.py:
```diff
@@ -0,0 +1,491 @@
+        )
+        relation_data = event.relation.data[self.model.app]
+        relation_data_dict = _load_relation_data(relation_data)
+        current_auth_conf = relation_data_dict.get("grafana_auth")
```
This is relation app data, not unit data. Under what circumstances would we expect a newly-joined relation to have data here?

saltiyazan replied:
Good point. Done :)

rbarry82 commented on lib/charms/grafana_auth_interface/v0/grafana_auth_interface.py:
```diff
@@ -0,0 +1,491 @@
+        relation_data_dict = _load_relation_data(relation_data)
+        current_auth_conf = relation_data_dict.get("grafana_auth")
+        if not current_auth_conf:
+            relation_data["grafana_auth"] = json.dumps(self._auth_conf)
```
This ends up with:

```python
{
  "grafana_auth": {
      "auth": {
         "auth_type_1": {... },
         "auth_type_2": { ... },
      },
  },
}
```
Which seems like one more layer than we really need.

saltiyazan replied:
Fixed

rbarry82 commented on lib/charms/grafana_auth_interface/v0/grafana_auth_interface.py:
```diff
@@ -0,0 +1,491 @@
+            return
+        if not self._app_relation_data_is_valid(grafana_url):
+            logger.warning("Relation data did not pass JSON Schema validation")
+            return
```
This feels a little tortured.

```python
relation_data = _load_relation_data(event.relation.data[event.app])

if not relation_data:
    return
    
if "grafana_url" not in relation_data or not self._app_relation_data_is_valid(self._app_relation_data_is_valid(relation_data.get("grafana_url"))...
   logger.warning(...)
```

But even that doesn't feel Pythonic, honestly.

Let's step back a bit. `_load_relation_data()` blindly loads all of the keys, not just the ones needed in either place (`grafana_url` or `grafana_auth` in `_relation_joined`). Operator Framework lazily loads data. Iterating through everything in the data bag, even the ones which aren't necessary, will mean a fair amount of RPC/API calls to Juju. It also means that any reader of this library needs to refer back to `_load_relation_data` frequently, which doesn't really appear to offer much over:

```python
grafana_url = json.loads(event.relation.data[event.app].get("grafana_url", ""))
if not grafana_url:
  return
```

Is there a reason this pattern was followed?

saltiyazan replied:
Agreed and fixed.

rbarry82 commented on lib/charms/grafana_auth_interface/v0/grafana_auth_interface.py:
```diff
@@ -0,0 +1,491 @@
+        )
+
+    @staticmethod
+    def _app_relation_data_is_valid(app_relation_data) -> bool:
```
Docstring is great. The method name reads (to me) more naturally and less verbosely as `validates_requirer_schema`, which tells at a glance what it actually does. I'm not sure why this is a `@staticmethod` instead of a top-level method or a normal instance method which the `event` (or the decoded JSON) could be passed to, though.

saltiyazan replied:
I changed it to inline validation :)

rbarry82 commented on lib/charms/grafana_auth_interface/v0/grafana_auth_interface.py:
```diff
@@ -0,0 +1,491 @@
+        Returns:
+            bool: True/False depending on whether the configuration follows the json schema.
+        """
+        relation_data = {"application-data" : app_relation_data}
```
This could be inlined in the method call.

saltiyazan replied:
Fixed

rbarry82 commented on lib/charms/grafana_auth_interface/v0/grafana_auth_interface.py:
```diff
@@ -0,0 +1,491 @@
+            validate(instance=relation_data, schema=REQUIRER_JSON_SCHEMA)
+            return True
+        except exceptions.ValidationError:
+            return False
```
We should log some information about _why_ it didn't validate.

saltiyazan replied:
Fixed.

rbarry82 commented on lib/charms/grafana_auth_interface/v0/grafana_auth_interface.py:
```diff
@@ -0,0 +1,491 @@
+            logger.warning("No authentication config")
+            return
+        if not self._app_relation_data_is_valid(auth_conf):
+            logger.warning("Relation data did not pass JSON Schema validation")
```
Same comments as the other side, really.

rbarry82 commented on lib/charms/grafana_auth_interface/v0/grafana_auth_interface.py:
```diff
@@ -0,0 +1,491 @@
+
+    def restore(self, snapshot: dict):
+        """Restores snapshot."""
+        self.auth_conf = snapshot["auth_conf"]
```
Please use [this](https://github.com/canonical/grafana-k8s-operator/blob/main/lib/charms/grafana_k8s/v0/grafana_dashboard.py#L631). snapshotted/restored `dict`s will come out as `StoredDict`, with nested `list`, `dict`, etc the same way. This makes comparisons with "native" python problematic, and `Stored*` will not JSON encode/decode, YAML load/dump -- I don't expect the Grafana charm to do any JSON/YAML manipulation to this, but if this library gets some of the utility extracted, other charms probably will.

rbarry82 commented on tests/charms/grafana_auth_interface/test_charm.py:
```diff
@@ -0,0 +1,370 @@
+        f"{CHARM_LIB_PATH}.GrafanaAuthRequirerCharmEvents.grafana_auth_config_available",
+        new_callable=PropertyMock,
+    )
+    def test_given_grafana_auth_conf_not_in_relation_databag_and_unit_is_leader_when_grafana_auth_relation_changed_then_grafana_auth_config_available_event_is_not_emitted(
```
These test names are insanely verbose. If a time comes where the is a test failure, anyone fixing it would need to actually edit the code and find it anyway. Do we need test names which are 162 characters long, without even including `def` or leading whitespace?

Docstrings or comments explaining what the test is intending to exercise would be a lot more readable. It's a fair assumption that some people will literally need to scroll to the side in their editor to finish seeing what the test is supposed to do.

saltiyazan replied:
We follow this naming convention in our charms, I'd be happy to agree on another convention to follow in this lib.

sed-i replied:
I also like to practice "loose Gherkin" ([example](https://github.com/canonical/alertmanager-k8s-operator/blob/fb9ad0012e125eac56049efe002bcf35839db1e9/tests/unit/test_external_url.py#L68)) but I agree with @rbarry82 here.

rbarry82 commented on lib/charms/grafana_auth_interface/v0/grafana_auth_interface.py:
```diff
@@ -0,0 +1,491 @@
+                    "proxy": {
+                        "enabled": True,
+                        "header_name": "some-header",
+                        "header_property": "some-header-property",
```
This really deserves a better example. Only `username` or `email` are allowed, and an indication that "this is the value which `header_name` will be mapped to would help. For that matter, putting a link to the complete [spec from Grafana](https://grafana.com/docs/grafana/latest/setup-grafana/configure-security/configure-authentication/auth-proxy/) would avoid some footguns

rbarry82 commented:
I'd love to have this live in the Grafana repo for now, at least until some kind of reusable pattern can be extracted.

I have three big worries about the current PR. The unit tests don't use `ops.testing` at all. `ops.testing.Harness` is not a perfect 1:1 mapping to the way OF actually behaves, but it's a much, much stronger guarantee than any of us simulating how we think it's supposed to behave by mocking things. Unit tests will never be perfect, but smashing it together with `Mock` doesn't have any intersection with the way OF may actually behave.

Secondly, this library doesn't "feel" right. I know that's subjective, but the `jsonschema` isn't actually validating header values which mean anything. It would be annoying to write a `pattern` which simulates an enum for a string with bare values and a regexp OR, but it would be some measure of sanity. Loading the entire app data bag (OF [lazily loads actual content](https://github.com/canonical/operator/blob/ac7b9f0944d7a313234b43ebda95df483dd17329/ops/model.py#L511), and iterating every key just to get one will lead to a `juju relation-get` for every single key, whether it's used or not). The entire pattern around `_load_relation_data` could just be dropped, since OF _does_ preload the keys, and "if a key exists, see if `json.loads(some_relation.data[some.app].get("key", ""))` is falsy" does the same thing with less churn on the backend.

Finally, the current implementation is in an uncanny valley where it's both too tightly and too loosely coupled to a particular implementation. It is `grafana-auth-interface`, but all of the keys are prepended with `grafana_` anyway, instead of simply being `url` or `auth`, and there's no default relation interface. `GrafanaAuthProvides` doesn't give any feedback to charm authors about what it expects. They need to either read the code (assuming they're familiar with `jsonschema`) or reverse engineer it, because `**kwargs` could be anything. 

The handling of validation is inverted from what I would expect. Specifically, it seems that `GrafanaAuthProvides` should be a parent class which handles common operations. From there, additional classes (in this case, `GrafanaAuthProxyProvider`, since that's the implemented case) could and should provide docstrings in the constructor which provide some hint to authors about what to pass instead of just `**kwargs`. Additionally, it's very hard to see why the requirer is doing the validation for an object which the provider has 100% of the information for, and leads to a bad admin experience where, as currently written, the grafana charm emits a `logger.warning` message that relation data didn't pass validation. From which relation (if there's more than one)? Which key? The provider has all of the information needed to validate.

As a broader comment, `jsonschema` is an amazing backstop, but it's not a good reason from my POV not to _try_ to sanity check it in the Python code itself. If an argument is clearly invalid, a single `logger.warning|error` about that specific value is much less "churn" than output from `jsonschema.validate`, and much more helpful to authors.

The library also does not seem to account for "what if more than one Grafana is related"? If we take this as a skeleton to base a more "generic" auth library on, some oauth/oidc/auth_proxy may have many consumers. Even if it's just Grafana, it may have more than one `grafana_url`

The spec discussed nesting, but from the PR, seeing a structure like:

```
<grafana_auth:interface>: {
  "auth": {
      "auth_proxy": { ... },
      "auth_generic_oauth": { ... },
      "auth_azuread": { ... },
  },
  "clients": {
      "grafana": [ ... ],
  },
}
```

Then:

```python
class GrafanaAuthProvider(Object):
  ...
  def _build_configuration(self, relation: Relation, auth_type: str, **kwargs):
      if not self._charm.is_leader():
          return
         
      # don't stomp on configuration from other providers in the same relation/app
      current_configuration = relation.data[self._charm.app].get("auth", {})
      current_configuration.update({auth_type: { k, v for k, v in kwargs.items() }})
      return current_configuration
      
class GrafanaAuthProxyProvider(GrafanaAuthProvider):
    def __init__(self, charm: CharmBase, relationship_name: str = "grafana-auth", header_name: str = "X-WEBAUTH-USER", ...)
```

This would avoid the "double nesting" of relation data, multiple clients, and a clearly defined class structure where each "auth type" could verify/sanitize inputs in Python and _then_ validate that it validates jsonschema, finally passing it over relation data so it's clear (if there's a validation failure) where a failure came from, and hopefully what the value was.

The sheer size of `jsonschema` definitions means that expanding this out to the 9 types (ignoring grafana.com, etc) which we may someday add means that the file would be large, but at least it would be well-structured, and if we ever got to that point, breaking out the "base class" to be injected into each separately or required as a lib (since the functionality is quite small if we drop the assumptions it currently has) would be simpler.

I know this is a large comment, and I'm happy to submit a PR which implements most of it. The current state of the code looks like a "scrap and rewrite" when (not if) we are requested to extend this to support something other than auth.proxy, and doing that now keeps us all on the same page.

saltiyazan committed:
Addresses review comments
Removes app level relation data check in relation joined handler
Removes grafana-auth and grafana-url keys
Removes _load_relation_data()
Removes _app_relation_data_is_valid and validates data inline
Adds missing log messages
Fixes unit tests

rbarry82 commented:
Already looks nicer 👍 

I don't want to step on your work (git branching isn't _that_ good), but if you're open to it, I'd be happy to rework the unit tests to use `ops.testing` instead of `Mock` and take a run at reshuffling the class structure to pull "common" functionality in to a base class and make `GrafanaProxyAuthProvider` (or `GrafanaAuthProxyProvider`, whichever) into something which more-or-less only handles the `auth.proxy` parts (default args in the constructor, etc) as a means to make this easier to extend and a guidepost for anyone who wants to add new auth types (ala an interface -- these are the things you must implement to add a new type)

saltiyazan commented:
> I'd love to have this live in the Grafana repo for now, at least until some kind of reusable pattern can be extracted.
> 
> I have three big worries about the current PR. The unit tests don't use `ops.testing` at all. `ops.testing.Harness` is not a perfect 1:1 mapping to the way OF actually behaves, but it's a much, much stronger guarantee than any of us simulating how we think it's supposed to behave by mocking things. Unit tests will never be perfect, but smashing it together with `Mock` doesn't have any intersection with the way OF may actually behave.
> 
> Secondly, this library doesn't "feel" right. I know that's subjective, but the `jsonschema` isn't actually validating header values which mean anything. It would be annoying to write a `pattern` which simulates an enum for a string with bare values and a regexp OR, but it would be some measure of sanity. Loading the entire app data bag (OF [lazily loads actual content](https://github.com/canonical/operator/blob/ac7b9f0944d7a313234b43ebda95df483dd17329/ops/model.py#L511), and iterating every key just to get one will lead to a `juju relation-get` for every single key, whether it's used or not). The entire pattern around `_load_relation_data` could just be dropped, since OF _does_ preload the keys, and "if a key exists, see if `json.loads(some_relation.data[some.app].get("key", ""))` is falsy" does the same thing with less churn on the backend.
> 
> Finally, the current implementation is in an uncanny valley where it's both too tightly and too loosely coupled to a particular implementation. It is `grafana-auth-interface`, but all of the keys are prepended with `grafana_` anyway, instead of simply being `url` or `auth`, and there's no default relation interface. `GrafanaAuthProvides` doesn't give any feedback to charm authors about what it expects. They need to either read the code (assuming they're familiar with `jsonschema`) or reverse engineer it, because `**kwargs` could be anything.
> 
> The handling of validation is inverted from what I would expect. Specifically, it seems that `GrafanaAuthProvides` should be a parent class which handles common operations. From there, additional classes (in this case, `GrafanaAuthProxyProvider`, since that's the implemented case) could and should provide docstrings in the constructor which provide some hint to authors about what to pass instead of just `**kwargs`. Additionally, it's very hard to see why the requirer is doing the validation for an object which the provider has 100% of the information for, and leads to a bad admin experience where, as currently written, the grafana charm emits a `logger.warning` message that relation data didn't pass validation. From which relation (if there's more than one)? Which key? The provider has all of the information needed to validate.
> 
> As a broader comment, `jsonschema` is an amazing backstop, but it's not a good reason from my POV not to _try_ to sanity check it in the Python code itself. If an argument is clearly invalid, a single `logger.warning|error` about that specific value is much less "churn" than output from `jsonschema.validate`, and much more helpful to authors.
> 
> The library also does not seem to account for "what if more than one Grafana is related"? If we take this as a skeleton to base a more "generic" auth library on, some oauth/oidc/auth_proxy may have many consumers. Even if it's just Grafana, it may have more than one `grafana_url`
> 
> The spec discussed nesting, but from the PR, seeing a structure like:
> 
> ```
> <grafana_auth:interface>: {
>   "auth": {
>       "auth_proxy": { ... },
>       "auth_generic_oauth": { ... },
>       "auth_azuread": { ... },
>   },
>   "clients": {
>       "grafana": [ ... ],
>   },
> }
> ```
> 
> Then:
> 
> ```python
> class GrafanaAuthProvider(Object):
>   ...
>   def _build_configuration(self, relation: Relation, auth_type: str, **kwargs):
>       if not self._charm.is_leader():
>           return
>          
>       # don't stomp on configuration from other providers in the same relation/app
>       current_configuration = relation.data[self._charm.app].get("auth", {})
>       current_configuration.update({auth_type: { k, v for k, v in kwargs.items() }})
>       return current_configuration
>       
> class GrafanaAuthProxyProvider(GrafanaAuthProvider):
>     def __init__(self, charm: CharmBase, relationship_name: str = "grafana-auth", header_name: str = "X-WEBAUTH-USER", ...)
> ```
> 
> This would avoid the "double nesting" of relation data, multiple clients, and a clearly defined class structure where each "auth type" could verify/sanitize inputs in Python and _then_ validate that it validates jsonschema, finally passing it over relation data so it's clear (if there's a validation failure) where a failure came from, and hopefully what the value was.
> 
> The sheer size of `jsonschema` definitions means that expanding this out to the 9 types (ignoring grafana.com, etc) which we may someday add means that the file would be large, but at least it would be well-structured, and if we ever got to that point, breaking out the "base class" to be injected into each separately or required as a lib (since the functionality is quite small if we drop the assumptions it currently has) would be simpler.
> 
> I know this is a large comment, and I'm happy to submit a PR which implements most of it. The current state of the code looks like a "scrap and rewrite" when (not if) we are requested to extend this to support something other than auth.proxy, and doing that now keeps us all on the same page.

Thank you for comments and thorough analysis of the PR, it definitely helps me understanding the framework, charm relations and libraries in general :)
I have addressed most of your review comments, it for sure made the code cleaner.

For your worries on the PR:
I have addressed some of them like `Double nesting`, but I think let's have a live discussion to decide on the matters that you have mentioned to be sure that we are on the same page regarding the design of this library.

Note: I changed this library to draft so I work on new unit tests that use `ops.testing`.

sed-i commented on CONTRIBUTING.md:
```diff
@@ -0,0 +1,28 @@
+    virtualenv -p python3 venv
+    source venv/bin/activate
+
+## Testing
```
We use tox, we do not use venv directly anymore.

```suggestion
```

saltiyazan replied:
Fixed

sed-i commented on README.md:
```diff
@@ -0,0 +1,25 @@
+
+## Description
+
+This project contains libraries for the grafana-auth relationship. It contains both the provider and the requirer side of them.
```
```suggestion
This project contains libraries for the grafana-auth relationship. It contains both the provider and the requirer sides.
```

saltiyazan replied:
Fixed

sed-i commented on README.md:
```diff
@@ -0,0 +1,25 @@
+
+This project contains libraries for the grafana-auth relationship. It contains both the provider and the requirer side of them.
+
+> Warning: The charm located here is a placeholder charm and should not be packed nor deployed.
```
Could / should it be used as the tester charm? ([example](https://github.com/canonical/observability-libs/blob/main/tests/integration/test_compute_resources_patch.py#L44))

sed-i commented on metadata.yaml:
```diff
@@ -0,0 +1,16 @@
+
+name: grafana-auth-interface
+display-name: |
+  grafana-auth-interface
```
display-name is deprecated ([ref](https://juju.is/docs/sdk/metadata-reference) doesn't mention it explicitly but it's gone).

```suggestion
```

saltiyazan replied:
Fixed

sed-i commented on metadata.yaml:
```diff
@@ -0,0 +1,16 @@
+resources:
+  placeholder-image:
+    type: oci-image
+    description: Placeholder for OCI image
```
This can be omitted ([example](https://github.com/canonical/prometheus-scrape-config-k8s-operator/blob/main/metadata.yaml)).
```suggestion
```

saltiyazan replied:
Fixed

sed-i commented on tests/charms/grafana_auth_interface/test_charm.py:
```diff
@@ -0,0 +1,342 @@
+#!/usr/bin/env python3
```
I think our convention is to place unit tests under `tests/unit`.

saltiyazan replied:
Fixed

sed-i commented on lib/charms/grafana_auth_interface/v0/grafana_auth_interface.py:
```diff
@@ -0,0 +1,416 @@
+# See LICENSE file for licensing details.
+
+"""grafana-auth interface library.
+This library implements the grafana-auth relation interface, it contains the Requires and Provides classes for handling
```
```suggestion
This library implements the grafana-auth relation interface, it contains the Requirer and Provider classes for handling
```

saltiyazan replied:
Fixed

sed-i commented on lib/charms/grafana_auth_interface/v0/grafana_auth_interface.py:
```diff
@@ -0,0 +1,416 @@
+            charm.on[relationship_name].relation_changed, self._on_auth_relation_changed
+        )
+
+    def _build_conf_dict(self, auth_type, **kwargs) -> dict:
```
Would it be easier if instead `**kwargs` the user would:
```python
from charms.grafana_auth_interface.v0.grafana_auth_interface import ProxyAuth


my_proxy_auth = ProxyAuth(what=ever)
self.grafana_auth = GrafanaAuthProvides(self, method=my_proxy_auth)
```

saltiyazan replied:
Fixed (will be using inheritance :) )

sed-i commented on lib/charms/grafana_auth_interface/v0/grafana_auth_interface.py:
```diff
@@ -0,0 +1,416 @@
+        """
+        return {auth_type: { key:value for key, value in kwargs.items()}}
+
+    def _set_auth_config_in_relation_data(self, event: RelationJoinedEvent) -> None:
```
Would this eventually be also called on leader-elected?
If the relation event takes place while the leader is unreachable but before a new one was elected, this reldata would never be set. ([example](https://github.com/canonical/prometheus-k8s-operator/blob/e75eb0b1a7669c945dabb6b4ba8cfe583697533b/lib/charms/prometheus_k8s/v0/prometheus_scrape.py#L1517))

saltiyazan replied:
```suggestion
    def _set_auth_config_in_relation_data(self, event: RelationJoinedEvent) -> None:
        if not self._charm.unit.is_leader():
            event.defer()
            return
```
Would adding `event.defer()` be a good solution instead of calling the function on both events?

saltiyazan replied:
I realised that it doesn't make sense to use `event.defer()` here, so the handler will run on `leader elected` as you suggested. Fixed :)

saltiyazan committed:
Addresses review comments
Adds a subclass for proxy auth mode
Adds refresh events
Updates unit tests

saltiyazan committed:
Fixes relation data
Fixes application data in relation data to use the correct application as a key

saltiyazan committed:
Adds unit test cases
Adds unit test cases that use dummy charms to test using harness

saltiyazan committed:
Sets data in relation databag on leader-elected event
Fixes related test cases

saltiyazan committed:
Fixes tox static analysis issues

saltiyazan committed:
Adds initial integration testing

sanchezfdezjavier commented on src/charm.py:
```diff
@@ -0,0 +1,14 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Contains a palceholder dummy charm."""
```
```suggestion
"""Contains a placeholder dummy charm."""
```

saltiyazan replied:
fixed

sanchezfdezjavier commented on tests/integration/auth_requirer_tester/src/charm.py:
```diff
@@ -0,0 +1,38 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""A charm to functionally test the auth provider in gafana-auth library."""
```
```suggestion
"""A charm to functionally test the auth provider in grafana-auth library."""
```

saltiyazan replied:
fixed

sanchezfdezjavier commented on tests/integration/auth_provider_tester/src/charm.py:
```diff
@@ -0,0 +1,39 @@
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""A charm to functionally test the auth provider in gafana-auth library."""
```
```suggestion
"""A charm to functionally test the auth provider in grafana-auth library."""
```

saltiyazan replied:
fixed

sanchezfdezjavier commented on tests/integration/test_grafana_auth.py:
```diff
@@ -0,0 +1,96 @@
+            trust=True,
+        )
+
+    async def test_relate_and_wait_for_idle(self, ops_test):
```
Missing fixture?
```suggestion
    async def test_relate_and_wait_for_idle(self, ops_test, setup):
```

saltiyazan replied:
fixed

sanchezfdezjavier commented on lib/charms/grafana_auth_interface/v0/grafana_auth_interface.py:
```diff
@@ -0,0 +1,585 @@
+    if isinstance(obj, StoredList):
+        return list(map(_type_convert_stored, obj))
+    elif isinstance(obj, StoredDict):
+        rdict = {}  # type: Dict[Any, Any]
+        for k in obj.keys():
+            rdict[k] = _type_convert_stored(obj[k])
+        return rdict
```
A bit cleaner?
```suggestion
        return {k: _type_convert_stored(obj[k]) for k in obj.keys()}
```

saltiyazan replied:
I agree. Fixed

sanchezfdezjavier commented on lib/charms/grafana_auth_interface/v0/grafana_auth_interface.py:
```diff
@@ -0,0 +1,585 @@
+        relation_data = relation.data[self._charm.app]
+        relation_data[AUTH] = json.dumps(self._auth_config)
+
+    def _get_urls_from_relation_data(self,event: Union[PebbleReadyEvent, RelationChangedEvent]) -> None:
```
```suggestion
    def _get_urls_from_relation_data(self, event: Union[PebbleReadyEvent, RelationChangedEvent]) -> None:
```

saltiyazan replied:
Fixed all lint issues in the lib file :)

sanchezfdezjavier commented on lib/charms/grafana_auth_interface/v0/grafana_auth_interface.py:
```diff
@@ -0,0 +1,585 @@
+            relation_id = relation.id
+        )
+
+    def _validate_auth_config_json_schema(self) -> bool:
+        """Validates authentication configuration using json schemas
+        Returns:
+            bool: Whether the configuration is valid or not based on the json schema.
+        """
+        return False
```
What is the purpose of this method?

saltiyazan replied:
This function is implemented in the child class and I had to avoid using `abc` so I could raise `NotImplementedError` here. Do you think that would give more feedback?

saltiyazan committed:
Addresses review comments and fixes integration test
Fixes the path to the lib
Fixes typos

saltiyazan committed:
Fixes lint issues

rbarry82 reviewed (commented):
This looks great now (I know it's still `[DRAFT]`). How do you feel about pulling it into the Grafana repo directly?

simonaronsson commented:
> This looks great now (I know it's still `[DRAFT]`). How do you feel about pulling it into the Grafana repo directly?

I was just about to say the same. Consider closing this PR, move the lib to the grafana-k8s-operator repo and then archive this repo. 👍🏼
